### PR TITLE
Add tests for `mongodump` and `mongorestore`

### LINKED ISSUE #149: Add tests for `mongorestore` and `mongodump`
We should run `mongorestore` and `mongodump` with various flags to ensure we stay compatible with FerretDB. Let's add a new set of tests that would run binaries, not use their tests.

* https://www.mongodb.com/docs/database-tools/mongorestore/
* https://www.mongodb.com/docs/database-tools/mongodump/
* https://github.com/mongodb/mongo-tools
* https://github.com/mcampo2/mongodb-sample-databases

### PR DESCRIPTION
**Pull request by noisersup**
Closes #149.

We can't compare BSON files after all.

### CODE REVIEW

**Comment by noisersup** [comment]
I've definitely overcomplicated the solution, thanks! :slightly_smiling_face:

**Comment by chilagrow** [comment]
Thank you for the in depth explanation :hugs: 
> Dump file created by FerretDB differs from the mongodb's one. It's because of a different order of documents in collections (NOTE: do we have any plans to fix that?)

Let's create an issue so we can track it. Then we can decide how we prioritise it.

> **To make sure that the dump file doesn't contain invalid data we can:**

I don't quite understand what we can achieve from this. I think this was probably introduced because _something_ wasn't working as expected? If that's the case we should create an issue and test for the specific case instead of _While using FerretDB_, we should have _while we execute specific command which causes invalid data_.

**Comment by noisersup** [comment]
> Would this ticket https://github.com/FerretDB/FerretDB/issues/457 fix it? It's about sorting

It looks like it will! Thanks for noticing that :)

> Is this trying to compare out of order documents in the collection? For example, every single document occures at the same index does this mean we get documents sorted by _id in both database and check sorted collections are the same?

This point was written before change about sorting by `_id`. Now documents in tests are sorted before a dump to make dumps the same.

File: tests/mongo-tools/dump_restore_test.go
Comment (rumyantseva) [inline tests/mongo-tools/dump_restore_test.go:37]: So, in the scope of the current task, we use only this dump. Are there plans to test more dumps in the future, or do we consider this one enough?

File: tests/mongo-tools/dump_restore_test.go
Comment (rumyantseva) [inline tests/mongo-tools/dump_restore_test.go:96]: Let's maybe add a comment why we ignore metadata.json files.

**Review by rumyantseva — COMMENTED** [review]
The changes look reasonable to me. 

Asked a couple of things for clarification.

File: tests/mongo-tools/dump_restore_test.go
Comment (noisersup) [reply tests/mongo-tools/dump_restore_test.go:37]: We needed this dump to test some large real-word dataset. I believe, we don't have any plans for other test cases

Comment (rumyantseva) [reply tests/mongo-tools/dump_restore_test.go:37]: Ok!

File: tests/mongo-tools/dump_restore_test.go
Comment (noisersup) [reply tests/mongo-tools/dump_restore_test.go:97]: done!

**Review by rumyantseva — APPROVED** [review]
LGTM. I see dance tests fail at the moment, but it looks relevant to yesterday's changes, not to the current PR.

**Review by chilagrow — APPROVED** [review]

**Comment by noisersup** [comment]
> Merging is blocked
Merging can be performed automatically with 2 approving reviews.

It looks like approving reviews are not counted after merging main to the branch

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by chilagrow — APPROVED** [review]

**Review by AlekSi — COMMENTED** [review]
We also should not put binary files into git repo. Let's add https://github.com/mcampo2/mongodb-sample-databases as a git submodule

File: tests/mongo-tools/helpers.go
Comment (noisersup) [reply tests/mongo-tools/helpers.go:134]: Yeah, that's right. I don't remember what was my reasoning behind that. Fixed and checked in logs if it matches the path

**Comment by AlekSi** [comment]
> It was added like that because the dump on the submodule is fully in the json format

https://github.com/mcampo2/mongodb-sample-databases/tree/master/dump/sample_geospatial

> I cannot see any advantages of using submodule in this case.

git submodule does not take repository space; it is a separate repository by definition.

File: tests/mongo-tools/helpers.go
Comment (noisersup) [reply tests/mongo-tools/helpers.go:153]: It's because we don't want to treat directories as files and compare their nonexisting "content". The `WalkDir` function should check for the files of this directory at some iteration.

File: docker-compose.yml
Comment (noisersup) [reply docker-compose.yml:59]: I thought placing all expected and actual dumps into one root directory of the project would be just better than having two additional top level directories

**Comment by noisersup** [comment]
> https://github.com/mcampo2/mongodb-sample-databases/tree/master/dump/sample_geospatial

Thanks!

> git submodule does not take repository space; it is a separate repository by definition.

If you want to run dance tests you still need to fetch whole submodule and as far as I recall there's no easy way to just fetch specific directory from submodule. So we'll need to fetch additional 1GB of data on every CI run

**Comment by AlekSi** [comment]
Correct – fetch, but not store. Fetching data from GitHub on GitHub's CI is fast. Working with a large repository with binary files is slow for both GitHub's CI and us.

**Comment by noisersup** [comment]
I see your point. Let me apply the changes I will re-request review asap to make it merged, thanks!

File: docker-compose.yml
Comment (AlekSi) [reply docker-compose.yml:59]: I think having `/dumps/actual` and `/dumps/expected` on both the host and inside the container would simplify things

File: tests/mongo-tools/helpers.go
Comment (noisersup) [reply tests/mongo-tools/helpers.go:132]: done

File: tests/mongo-tools/helpers.go
Comment (noisersup) [reply tests/mongo-tools/helpers.go]: done

File: docker-compose.yml
Comment (noisersup) [reply docker-compose.yml]: done

File: tests/mongo-tools/dump_restore_test.go
Comment (chilagrow) [inline tests/mongo-tools/dump_restore_test.go:90]: Do want to check that `metadata.json` exist in both localActualPath and localExpectedPath?

File: tests/mongo-tools/dump_restore_test.go
Comment (chilagrow) [inline tests/mongo-tools/dump_restore_test.go:86]: We don't need to do this comparison for this task, I think that was our discussion :hugs:

File: tests/mongo-tools/dump_restore_test.go
Comment (noisersup) [reply tests/mongo-tools/dump_restore_test.go:90]: *.metadata.json is the file containing index definitions and it's not required to perform a dump.
Yes, we could check at least if it exists, because it's actually created on every dump (with a different uuid field), although we don't check how `mongodump` works, but if it works with FerretDB so imo we cand just skip it as we don't even provide the index functionality

File: tests/mongo-tools/dump_restore_test.go
Comment (noisersup) [reply tests/mongo-tools/dump_restore_test.go:86]: I believe that discussion was focused more on testing if FerretDB omits any data from the external **source dump** generated by MongoDB. We've already merged the database state comparison with the past PRs, so I assumed it's still in a scope.

We aim to compare dumps created by FerretDB, but as binary files are... _just binary files_, the comparison of them is just a matter of saying "Yes they're equal" or "No, they're not". 

Comparing the database state under both of these dumps is just handful for the output purposes.

Let's maybe leave this conversation open to discuss it with everyone if needed :eyes:

**Review by w84thesun — APPROVED** [review]
Changes are looking reasonable to me. I thought there might be some sorting problems but it's working, I guess.

File: tests/mongo-tools/dump_restore_test.go
Comment (chilagrow) [reply tests/mongo-tools/dump_restore_test.go:90]: thanks for the explanation. Yes we ignore index for now. Didn't realise metadata was related to index :see_no_evil:

File: tests/mongo-tools/dump_restore_test.go
Comment (chilagrow) [reply tests/mongo-tools/dump_restore_test.go:86]: Yeah okie, let's keep it for now.

**Review by chilagrow — APPROVED** [review]

File: tests/mongo-tools/helpers.go
Comment (AlekSi) [reply tests/mongo-tools/helpers.go:165]: Then I guess that should be just
```go
if d.IsDir() {
 return nil
}
```

I don't understand how `WalkDir` could find non-existing directory

**Comment by AlekSi** [comment]
It is not ready for review

File: tests/mongo-tools/helpers.go
Comment (noisersup) [reply tests/mongo-tools/helpers.go:165]: I think I don't understand. Let's say we have 2 directories that we want to compare: A and B.
```
├── A/
│   ├── boo/
│   └── foo/
│       └── bar.txt
└── B/
    ├── boo/
    └── foo/
        └── bar.txt
```
Walkdir that is running on A/ even when A/boo/ is empty will go through it, so it's worth to check if the B/boo/ also exists.

**Comment by noisersup** [comment]
> It is not ready for review

Why?

**Comment by AlekSi** [comment]
GitHub keeps eating the large comments I wrote with a lousy internet connection; only two went through 😞. See https://github.com/FerretDB/dance/pull/262/files#r1049605371 and https://github.com/FerretDB/dance/pull/262/files#r1049599247. Basically, something weird is going on if the code that could not be correct somehow passes tests. I'm taking a deeper look.

File: tests/mongo-tools/helpers.go
Comment (noisersup) [reply tests/mongo-tools/helpers.go:132]: First and foremost thanks for a wide explanation :pray:

> It is not clear why we specifically check that file exists, given that os.Open would return the same error.

This change was made only to make the output little bit more readable if file does not exist.

> I'm not sure what we do next. It seems that this code tries to compare files' content – but why, given that we already compared hashes exactly to prevent reading the whole files into memory?

This part of code is called only when checksums are different and it was added only to be more verbose about specific differences in files.

> And then why do we rewind only the first file but not the second?

Nice catch, it doesn't fail because it's not even called, as checksums are the same

File: tests/mongo-tools/helpers.go
Comment (chilagrow) [reply tests/mongo-tools/helpers.go:172]: Just so I understand, this is about testing _expected_ (_dir1_) having an empty directory but _actual_ (_dir2_) is not existing? I feel this isn't necessary since there is no files inside neither directory, we don't worry about if it exist or not.

What's more interesting is if _expected_ (_dir1_) have an empty directory but _actual_ (_dir2_) has files. But let's not go there in this task for now.

**Review by rumyantseva — COMMENTED** [review]
Could you please re-request review when the PR is ready?

**Review by w84thesun — COMMENTED** [review]
Please re-request when ready.

**Review by w84thesun — APPROVED** [review]
LGTM

**Review by chilagrow — APPROVED** [review]
Great!

**Comment by AlekSi** [comment]
@noisersup PTAL

**Review by AlekSi — APPROVED** [review]

File: tests/mongo-tools/dump_restore_test.go
Comment (noisersup) [inline tests/mongo-tools/dump_restore_test.go:85]: I think I might still don't understand why we are unable to generate same dumps. The "bootstrapping" approach was able to generate 2 same dumps as the order on the same database is preserved. Why we removed it?

File: tests/mongo-tools/dump_restore_test.go
Comment (noisersup) [inline tests/mongo-tools/dump_restore_test.go:102]: cool! :)

**Review by noisersup — COMMENTED** [review]
I cannot approve my PR, but LGTM. One question

File: tests/mongo-tools/dump_restore_test.go
Comment (AlekSi) [reply tests/mongo-tools/dump_restore_test.go:85]: Because it doesn't work for the second test with a slightly bigger database. The fact that the file was exactly the same for a smaller database was a happy accident

**Pull request merged**
